src/foam/dao/MongoDAO.java

```diff
@@ -0,0 +1,176 @@
+  }
+
+  private boolean isUserPassProvided(String username, String password) {
+    return ( ( username != null ) && ( ! username.isEmpty() ) ) &&
+            ( ( password != null ) && ( ! password.isEmpty() ) );
+  }
+
+  @Override
+  public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    if ( sink == null ) {
+      sink = new ListSink();
+    }
+
+    Sink         decorated = decorateSink_(sink, skip, limit, order, predicate);
+    Subscription sub       = new Subscription();
+
+    String collectionName = (String) x.get("collectionName");
+    String collectionClass = (String) x.get("collectionClass");
+    List props = getFObjectProperties(collectionClass, x);
+
+    if ( collectionName == null || collectionName.isEmpty() ) {
+      throw new IllegalArgumentException("Invalid collection name in context. Please provide a string name.");
+    } else if ( collectionClass == null || collectionClass.isEmpty() ) {
+      throw new IllegalArgumentException("Invalid collection class name in context. Please provide a string class name.");
+    }
+
+    MongoCollection<Document> collection = database.getCollection(collectionName);
+    MongoCursor<Document> cursor = collection.find().iterator();
+
+    try {
+      while ( cursor.hasNext() ) {
+        if ( sub.getDetached() ) break;
+
+        FObject obj = createFObject(collectionClass, props, cursor.next());
+
+        if ( ( predicate == null ) || predicate.f(obj) ) {
+          decorated.put(obj, sub);
+        }
+      }
+    } finally {
+      cursor.close();
+    }
+
+    decorated.eof();
+
+    return sink;
+  }
+
+  private List getFObjectProperties(String clsName, X x) {    
+    try {
+      Class cls = Class.forName(clsName);
+      FObject clsInstance = (FObject) x.create(cls);
+      List props = clsInstance.getClassInfo().getAxiomsByClass(foam.core.AbstractFObjectPropertyInfo.class);
+      
+      // Recursively handle nested FObjectProperties
+      for ( int i = 0 ; i < props.size() ; i++ ) {
+        props.addAll(getFObjectProperties(((PropertyInfo) props.get(i)).getPropertyType(), x));
+      }
+
+      return props;
+    } catch (Exception ex) {
+      throw new RuntimeException(clsName + " was not found.");
+    }
+  }
+
+  private FObject createFObject(String clsName, List props, Document d) {
+    JsonWriterSettings writerSettings = new JsonWriterSettings(JsonMode.SHELL, true);
+    String jsonStr = d.toJson(writerSettings);
+
+    // Trims initial `"_id" : ObjectId("[24 HEX Chars]"),`
+    jsonStr = "{ class: \"" + clsName + "\", " + 
+                 jsonStr.substring(MONGO_OBJECT_PREFIX_LENGTH, jsonStr.length() - 1) + 
+              " }";
+
+    Iterator i = props.iterator();
+
+    while ( i.hasNext() ) {
+      PropertyInfo prop = (PropertyInfo) i.next();
+      jsonStr = jsonStr.replace("\"" + prop.getName() + "\" : {", "\"" + prop.getName() + "\" : { class: \"" + prop.getPropertyType() + "\",");
```

Comment:
I don't think prop.getPropertyType() is necessary.  And it may lead to bugs.  When an FObjectProperty is serialized, the FObject it contains should serialize class: '' into the JSON and that's the class we should use when deserializing.

Consider this example

```
foam.CLASS({
  name: 'Abc',
  properties: ['a']
});

foam.CLASS({
  name: 'Def',
  extends: 'Abc',
  properties: ['b']
});

foam.CLASS({
  name: 'SomeClass',
  properties: [
  { class: 'FObjectProperty', of: 'Abc', name: 'foo' }
  ]
});
```

If I have an instance of SomeClass and I see foo = Def.create({ a: 1, b: 2 });  After saving and re-loading this instance from your DAO, I will instead get back foo == Abc.create({ a: 1 }), the fact that I had a Def instance stored will be lost.